design-proposals/mass_machine_type_transition.md

```diff
@@ -0,0 +1,265 @@
+﻿# Overview
+This proposal discusses a method to automate updating the machine type of VMs
+to the latest machine type version, if the VM has a machine type that is no
+longer compatible. Currently we can support manually changing the machine type of
+individual VMs, but we need an automated process that will be able to change the
+machine types of multiple (e.g. thousands of) VMs, limiting workload interruptions
+and preventing the user from having to manually update the machine type of every
+VM with a machine type version that is no longer supported. For example, CentOS
+Stream 9 will maintain compatibility with all `pc-q35-rhel8.x.x` machine types
+throughout its lifecycle. However, with the transition to CentOS Stream 10,
+compatibility with machine types prior to `pc-q35-rhel9.0.0` will not be maintained: 
+```
+$ cat /etc/centos-release 
+CentOS Stream release 9
+$ rpm -qa | grep qemu-kvm-core
+qemu-kvm-core-8.0.0-5.el9.x86_64
+$ /usr/libexec/qemu-kvm -machine ?
+Supported machines are:
+pc                   RHEL 7.6.0 PC (i440FX + PIIX, 1996) (alias of pc-i440fx-rhel7.6.0)
+pc-i440fx-rhel7.6.0  RHEL 7.6.0 PC (i440FX + PIIX, 1996) (default) (deprecated)
+q35                  RHEL-9.2.0 PC (Q35 + ICH9, 2009) (alias of pc-q35-rhel9.2.0)
+pc-q35-rhel9.2.0     RHEL-9.2.0 PC (Q35 + ICH9, 2009)
+pc-q35-rhel9.0.0     RHEL-9.0.0 PC (Q35 + ICH9, 2009)
+pc-q35-rhel8.6.0     RHEL-8.6.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.5.0     RHEL-8.5.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.4.0     RHEL-8.4.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.3.0     RHEL-8.3.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.2.0     RHEL-8.2.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.1.0     RHEL-8.1.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.0.0     RHEL-8.0.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel7.6.0     RHEL-7.6.0 PC (Q35 + ICH9, 2009) (deprecated)
+none                 empty machine
+```
+This means each workload must be updated to use a `pc-q35-rhel9.x.x` machine type.
+
+## Goals
+* Create an opt-in workflow that automates mass converting machine types of VMs
+	* if currently running, update VM machine type and add a label to alert the user the VM must be restarted for the change to take effect
```

Comment:
I think a fundamental issue is, that for whatever reason we are not resolving the cluster-wide machine type during VMI creation. Instead it gets added via a webhook during VM creation. This is problematic since it makes it impossible to differentiate explicit user-choices and cluster-wide decision points.

One option would be to rethink this behaviour. Nowadays the resolved machine type is also visible in the VMI status (also resolving libvirt aliases), and if we would not default the machine type, a restart would automatically pick up the latest type if unset on  the VM level, which would allow clear intent separation.

Reply:
I think the point of the mass machine type transition effort overall is to actually change the user requested value--as requested by the user, so in this case the approach in the design document makes sense.